mobilecoind-json/src/main.rs

```diff
@@ -0,0 +1,317 @@
+/// TODO: also add a POST that includes balance and memo 
+#[get("/monitors/<monitor_hex>/request-code/<index>")]
+fn request_code(
+    state: rocket::State<State>,
+    monitor_hex: String,
+    index: u64,
+) -> Result<Json<JsonRequestResponse>, String> {
+    let monitor_id =
+        hex::decode(monitor_hex).map_err(|err| format!("Failed to decode monitor hex: {}", err))?;
+
+    // Get our public address.
+    let mut req = mc_mobilecoind_api::GetPublicAddressRequest::new();
+    req.set_monitor_id(monitor_id.clone());
+    req.set_subaddress_index(index);
+
+    let resp = state
+        .mobilecoind_api_client
+        .get_public_address(&req)
+        .map_err(|err| format!("Failed getting public address: {}", err))?;
+
+    let public_address = resp.get_public_address().clone();
+
+    // Generate b58 code
+    let mut req = mc_mobilecoind_api::GetRequestCodeRequest::new();
+    req.set_receiver(public_address);
+    let resp = state
+        .mobilecoind_api_client
+        .get_request_code(&req)
+        .map_err(|err| format!("Failed getting request code: {}", err))?;
+
+    Ok(Json(JsonRequestResponse {
+        request_code: String::from(resp.get_b58_code()),
+    }))
+}
+
+#[derive(Deserialize)]
+struct JsonTransferRequest {
+    request_code: String,
+    amount: u64,
+}
+#[derive(Serialize)]
+struct JsonTransferResponse {
+    key_image: String,
+    tombstone: u64,
+}
+
+/// Performs a transfer from a monitor and subaddress. The target and amount are in the POST data.
+#[post(
+    "/monitors/<monitor_hex>/transfer/<index>",
+    format = "json",
+    data = "<transfer>"
+)]
+fn transfer(
+    state: rocket::State<State>,
+    monitor_hex: String,
+    index: u64,
+    transfer: Json<JsonTransferRequest>,
+) -> Result<Json<JsonTransferResponse>, String> {
+    let monitor_id =
+        hex::decode(monitor_hex).map_err(|err| format!("Failed to decode monitor hex: {}", err))?;
+
+    let mut req = mc_mobilecoind_api::ReadRequestCodeRequest::new();
+    req.set_b58_code(transfer.request_code.clone());
+    let resp = state
+        .mobilecoind_api_client
+        .read_request_code(&req)
+        .map_err(|err| format!("Failed reading request code: {}", err))?;
+    let public_address = resp.get_receiver();
+
+    let mut outlay = mc_mobilecoind_api::Outlay::new();
+    outlay.set_receiver(public_address.clone());
+    outlay.set_value(transfer.amount);
+
+    let mut req = mc_mobilecoind_api::SendPaymentRequest::new();
+    req.set_sender_monitor_id(monitor_id);
+    req.set_sender_subaddress(index);
+    req.set_outlay_list(RepeatedField::from_vec(vec![outlay]));
+
+    let resp = state
+        .mobilecoind_api_client
+        .send_payment(&req)
+        .map_err(|err| format!("Failed to send payment: {}", err))?;
+
+    let receipt = resp.get_sender_tx_receipt();
+    return Ok(Json(JsonTransferResponse {
+        key_image: hex::encode(receipt.get_key_image_list()[0].get_data()),
```

Comment:
This should return the entire receipt and let the client decide which data it wants to store.